src/WebJobs.Extensions.DurableTask/_CSharpLanguageHelpers.cs

```diff
@@ -0,0 +1,265 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+// This file defines several classes and methods that exist in .NET Core but not in .NET Standard 2.0.
+// They are defined here to enable certain C# features that otherwise require higher framework versions.
+// Redefining types in this way is a standard practice for libary authors that are forced to target .NET Standard 2.0.
+#if !NET31_OR_GREATER
+
+#pragma warning disable SA1649 // File name should match first type name
+#pragma warning disable SA1403 // File may only contain a single namespace
+#pragma warning disable SA1309 // Field names should not begin with underscore
+#pragma warning disable SA1101 // Prefix local calls with this
+#pragma warning disable SA1503 // Braces should not be omitted
+#pragma warning disable SA1201 // Elements should appear in the correct order
+#nullable enable
+
+using System.Runtime.CompilerServices;
+
+// Copied from https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/NullableAttributes.cs#L69
+namespace System.Diagnostics.CodeAnalysis
+{
+    internal sealed class NotNullWhenAttribute : Attribute
+    {
+        /// <summary>Initializes the attribute with the specified return value condition.</summary>
+        /// <param name="returnValue">
+        /// The return value condition. If the method returns this value, the associated parameter will not be null.
+        /// </param>
+        public NotNullWhenAttribute(bool returnValue) => this.ReturnValue = returnValue;
+
+        /// <summary>Gets the return value condition.</summary>
+        public bool ReturnValue { get; }
+    }
+}
+
+namespace System
+{
+    /// <summary>Represent a type can be used to index a collection either from the start or the end.</summary>
+    /// <remarks>
+    /// Index is used by the C# compiler to support the new index syntax.
+    /// <code>
+    /// int[] someArray = new int[5] { 1, 2, 3, 4, 5 } ;
+    /// int lastElement = someArray[^1]; // lastElement = 5
+    /// </code>
+    /// </remarks>
+    internal readonly struct Index : IEquatable<Index>
```

Comment:
Just out of curiosity, how did you know to copy this specific class? Was there a compiler error saying this was missing? I just don't see anywhere in the code where you call this directly, so I'm unsure how this specific implementation ended up in our _CSharpLanguageHelpers. Thanks!

Reply:
The C# compiler uses this type to implement the language features mentioned above in the comments. We're not using them directly. And yes, this struct is mentioned in the compiler error message but my knowledge of knowing to use it primarily came from researching how to use these newer language features when targeting .NET Standard 2.0.